src/dbms/replication_handler.cpp

```diff
@@ -180,28 +244,83 @@ auto ReplicationHandler::RegisterReplica(const memgraph::replication::Replicatio
   return {};
 }
 
+#ifdef MG_ENTERPRISE
+auto ReplicationHandler::RegisterMain(const memgraph::replication::ReplicationClientConfig &config)
+    -> utils::BasicResult<RegisterMainError> {
+  MG_ASSERT(dbms_handler_.ReplicationState().IsCoordinator(), "Only coordinator can register main instance!");
+
+  auto instance_client = dbms_handler_.ReplicationState().RegisterMain(config);
+  if (instance_client.HasError()) switch (instance_client.GetError()) {
+      case memgraph::replication::RegisterMainError::NOT_COORDINATOR:
+        MG_ASSERT(false, "Only coordinator can register main instance!");
+        return {};
+      case memgraph::replication::RegisterMainError::END_POINT_EXISTS:
+        return memgraph::dbms::RegisterMainError::END_POINT_EXISTS;
+      case memgraph::replication::RegisterMainError::COULD_NOT_BE_PERSISTED:
+        return memgraph::dbms::RegisterMainError::COULD_NOT_BE_PERSISTED;
+      case memgraph::replication::RegisterMainError::MAIN_ALREADY_EXISTS:
+        return memgraph::dbms::RegisterMainError::MAIN_ALREADY_EXISTS;
+      case memgraph::replication::RegisterMainError::SUCCESS:
+        break;
+    }
+
+  if (!allow_mt_repl && dbms_handler_.All().size() > 1) {
+    spdlog::warn("Multi-tenant replication is currently not supported!");
+  }
+
+  return {};
+}
+#endif
+
 auto ReplicationHandler::UnregisterReplica(std::string_view name) -> UnregisterReplicaResult {
   auto const replica_handler = [](RoleReplicaData const &) -> UnregisterReplicaResult {
-    return UnregisterReplicaResult::NOT_MAIN;
+    return UnregisterReplicaResult::IS_REPLICA;
   };
+
   auto const main_handler = [this, name](RoleMainData &mainData) -> UnregisterReplicaResult {
-    if (!dbms_handler_.ReplicationState().TryPersistUnregisterReplica(name)) {
+    if (!dbms_handler_.ReplicationState().TryPersistUnregisterReplicaOnMain(name)) {
       return UnregisterReplicaResult::COULD_NOT_BE_PERSISTED;
     }
+
     // Remove database specific clients
     dbms_handler_.ForEach([name](Database *db) {
       db->storage()->repl_storage_state_.replication_clients_.WithLock([&name](auto &clients) {
         std::erase_if(clients, [name](const auto &client) { return client->Name() == name; });
       });
     });
+
     // Remove instance level clients
     auto const n_unregistered =
         std::erase_if(mainData.registered_replicas_, [name](auto const &client) { return client.name_ == name; });
     return n_unregistered != 0 ? UnregisterReplicaResult::SUCCESS : UnregisterReplicaResult::CAN_NOT_UNREGISTER;
   };
 
+#ifdef MG_ENTERPRISE
+  auto const coordinator_handler =
+      [this, name](replication::RoleCoordinatorData &coordinatorData) -> UnregisterReplicaResult {
+    if (!dbms_handler_.ReplicationState().TryPersistUnregisterReplicaOnMain(name)) {
+      return UnregisterReplicaResult::COULD_NOT_BE_PERSISTED;
+    }
+
+    // Remove database specific clients
+    dbms_handler_.ForEach([name](Database *db) {
```

Comment:
Why does the coordinator care about storage clients?
Shouldn't it handle just the high-level?